File: server/lib/icecastStreamer.js
import { spawn } from 'child_process';
import { PassThrough } from 'stream';
import logger from '../utils/logger.js';
import { getFfmpegPath } from '../utils/utils.js';

class IcecastStreamer {
    constructor(config) {
        this.config = {
            host: config.host,
            port: config.port,
            password: config.password,
            mount: config.mount || '/radio.mp3',
            name: config.name || 'MRadio',
            description: config.description || 'MRadio Stream',
            genre: config.genre || 'Various',
            bitrate: config.bitrate || '128',
            sampleRate: config.sampleRate || '44100',
            channels: config.channels || '2'
        };
        
        this.ffmpegProcess = null;
        this.inputStream = null;
        this.isConnected = false;
        this.reconnectAttempts = 0;
        this.maxReconnectAttempts = 10;
        this.reconnectDelay = 5000;
        this.reconnectTimer = null;
        this.isReconnecting = false;
        this.buffer = [];
        this.bufferSize = 0;
        this.maxBufferSize = 1024 * 1024; // 1MB buffer
    }

    connect() {
        // Return existing connection promise if already connecting
        if (this.connectionPromise) {
            return this.connectionPromise;
        }

        this.connectionPromise = new Promise((resolve, reject) => {
            if (this.isConnected) {
                logger.info('Already connected to Icecast server');
                this.connectionPromise = null;
                return resolve();
            }

            if (this.isReconnecting) {
                logger.info('Already attempting to reconnect to Icecast server');
                this.connectionPromise = null;
                return resolve();
            }

            const icecastUrl = `icecast://source:${this.config.password}@${this.config.host}:${this.config.port}${this.config.mount}`;
            
            logger.info(`Connecting to Icecast server at ${this.config.host}:${this.config.port}${this.config.mount}`);

            // FFmpeg arguments for Icecast streaming with better error handling
            const ffmpegArgs = [
                '-hide_banner',
                '-loglevel', 'warning',
                // Input from stdin with re-sync on errors
                '-f', 'mp3',
                '-i', 'pipe:0',
                // Error recovery options
                '-err_detect', 'ignore_err',
                '-fflags', '+genpts+discardcorrupt',
                '-max_error_rate', '1.0',
                // Re-encode for consistent output
                '-acodec', 'libmp3lame',
                '-ab', `${this.config.bitrate}k`,
                '-ar', this.config.sampleRate,
                '-ac', this.config.channels,
                // Metadata
                '-metadata', `title="${this.config.name}"`,
                '-metadata', `artist="${this.config.description}"`,
                '-metadata', `genre="${this.config.genre}"`,
                // Output format with reconnect options
                '-f', 'mp3',
                '-reconnect', '1',
                '-reconnect_at_eof', '1',
                '-reconnect_streamed', '1',
                '-reconnect_delay_max', '5',
                // Icecast specific settings
                '-content_type', 'audio/mpeg',
                '-ice_name', this.config.name,
                '-ice_description', this.config.description,
                '-ice_genre', this.config.genre,
                '-ice_public', '1',
                // Output to Icecast
                icecastUrl
            ];

            this.ffmpegProcess = spawn(getFfmpegPath(), ffmpegArgs, {
                windowsHide: true
            });

            // Create input stream with high water mark for better buffering
            this.inputStream = new PassThrough({ highWaterMark: 64 * 1024 });
            this.inputStream.pipe(this.ffmpegProcess.stdin);

            // Flush any buffered data
            this.flushBuffer();

            // Handle FFmpeg stderr for debugging
            this.ffmpegProcess.stderr.on('data', (data) => {
                const message = data.toString();
                
                // Ignore certain non-critical errors during track transitions
                if (message.includes('Header missing') || 
                    message.includes('Invalid data found when processing input')) {
                    logger.debug(`FFmpeg warning (ignored during transition): ${message}`);
                    return;
                }
                
                if (message.includes('error') || message.includes('Error')) {
                    // Don't treat HTTP 403 as a critical error if we're already connected
                    if (message.includes('403 Forbidden') && this.isConnected) {
                        logger.debug('Received 403 during active connection, ignoring');
                        return;
                    }
                    logger.error(`FFmpeg Icecast error: ${message}`);
                    if (!this.isReconnecting) {
                        this.handleConnectionError();
                    }
                } else if (message.includes('Connected')) {
                    logger.info('Successfully connected to Icecast server');
                    this.isConnected = true;
                    this.isReconnecting = false;
                    this.reconnectAttempts = 0;
                    resolve();
                } else {
                    logger.debug(`FFmpeg Icecast: ${message}`);
                }
            });

            this.ffmpegProcess.on('error', (error) => {
                logger.error('FFmpeg process error:', error);
                if (!this.isReconnecting) {
                    this.handleConnectionError();
                }
                if (!this.isConnected) {
                    reject(error);
                }
            });

            this.ffmpegProcess.on('close', (code) => {
                if (code !== 0 && code !== null) {
                    logger.error(`FFmpeg process exited with code ${code}`);
                }
                this.isConnected = false;
                if (!this.isReconnecting) {
                    this.handleConnectionError();
                }
            });

            // Consider connected after a short delay if no errors
            setTimeout(() => {
                if (this.ffmpegProcess && !this.ffmpegProcess.killed) {
                    this.isConnected = true;
                    this.isReconnecting = false;
                    this.reconnectAttempts = 0;
                    logger.info('FFmpeg Icecast streaming initialized');
                    this.connectionPromise = null;
                    resolve();
                }
            }, 2000);
        }).finally(() => {
            // Clear the promise reference when done
            this.connectionPromise = null;
        });

        return this.connectionPromise;
    }

    handleConnectionError() {
        if (this.isReconnecting) {
            return;
        }
        
        this.isConnected = false;
        
        if (this.reconnectAttempts < this.maxReconnectAttempts) {
            this.scheduleReconnect();
        } else {
            logger.error('Max reconnection attempts reached for Icecast');
        }
    }

    scheduleReconnect() {
        if (this.isReconnecting) {
            return;
        }
        
        this.isReconnecting = true;
        this.reconnectAttempts++;
        const delay = Math.min(this.reconnectDelay * this.reconnectAttempts, 30000);
        
        logger.info(`Scheduling Icecast reconnection attempt ${this.reconnectAttempts}/${this.maxReconnectAttempts} in ${delay}ms`);
        
        // Clear any existing reconnect timer
        if (this.reconnectTimer) {
            clearTimeout(this.reconnectTimer);
        }
        
        this.reconnectTimer = setTimeout(() => {
            this.isReconnecting = false;
            this.reconnectTimer = null;
            
            // Clean up existing connection before reconnecting
            this.cleanup();
            
            this.connect().catch(err => {
                logger.error('Icecast reconnection failed:', err);
                this.isReconnecting = false;
            });
        }, delay);
    }

    flushBuffer() {
        if (this.buffer.length > 0 && this.inputStream && !this.inputStream.destroyed) {
            try {
                for (const chunk of this.buffer) {
                    this.inputStream.write(chunk);
                }
                this.buffer = [];
                this.bufferSize = 0;
                logger.debug('Flushed audio buffer to stream');
            } catch (error) {
                logger.error('Error flushing buffer:', error);
            }
        }
    }

    write(chunk) {
        if (!chunk || chunk.length === 0) return;

        // If we're connected and have a valid stream, write directly
        if (this.isConnected && this.inputStream && !this.inputStream.destroyed) {
            try {
                // Flush any buffered data first
                if (this.buffer.length > 0) {
                    this.flushBuffer();
                }
                
                const success = this.inputStream.write(chunk);
                if (!success) {
                    // Handle backpressure
                    this.inputStream.once('drain', () => {
                        logger.debug('Icecast stream drained');
                    });
                }
            } catch (error) {
                logger.error('Error writing to Icecast stream:', error);
                // Buffer the data instead of losing it
                this.addToBuffer(chunk);
                if (!this.isReconnecting) {
                    this.handleConnectionError();
                }
            }
        } else {
            // Buffer data while not connected
            this.addToBuffer(chunk);
            
            // Try to reconnect if not already attempting
            if (!this.isConnected && !this.isReconnecting && this.reconnectAttempts === 0) {
                this.connect().catch(err => {
                    logger.error('Failed to connect to Icecast:', err);
                });
            }
        }
    }

    addToBuffer(chunk) {
        // Add to buffer with size limit
        if (this.bufferSize + chunk.length <= this.maxBufferSize) {
            this.buffer.push(chunk);
            this.bufferSize += chunk.length;
        } else {
            // Remove oldest chunks to make room
            while (this.bufferSize + chunk.length > this.maxBufferSize && this.buffer.length > 0) {
                const removed = this.buffer.shift();
                this.bufferSize -= removed.length;
            }
            this.buffer.push(chunk);
            this.bufferSize += chunk.length;
        }
    }

    cleanup() {
        if (this.inputStream) {
            try {
                this.inputStream.end();
                this.inputStream.destroy();
            } catch (error) {
                logger.debug('Error cleaning up input stream:', error);
            }
            this.inputStream = null;
        }

        if (this.ffmpegProcess) {
            try {
                this.ffmpegProcess.stdin.end();
                this.ffmpegProcess.kill('SIGTERM');
            } catch (error) {
                logger.debug('Error cleaning up FFmpeg process:', error);
            }
            this.ffmpegProcess = null;
        }
    }

    disconnect() {
        // Clear reconnect timer
        if (this.reconnectTimer) {
            clearTimeout(this.reconnectTimer);
            this.reconnectTimer = null;
        }
        
        this.cleanup();
        
        this.isConnected = false;
        this.isReconnecting = false;
        this.reconnectAttempts = 0;
        this.buffer = [];
        this.bufferSize = 0;
        
        logger.info('Disconnected from Icecast server');
    }

    getStatus() {
        return {
            connected: this.isConnected,
            reconnecting: this.isReconnecting,
            reconnectAttempts: this.reconnectAttempts,
            bufferSize: this.bufferSize,
            config: {
                host: this.config.host,
                port: this.config.port,
                mount: this.config.mount,
                name: this.config.name
            }
        };
    }
}

export default IcecastStreamer;

File: app/streaming/icecast_streamer.py
import asyncio
import subprocess
from app.core import logger
from app.core.utils import get_ffmpeg_path

class IcecastStreamer:
    def __init__(self, config: dict):
        self.config = {
            'host': config.get('host'),
            'port': config.get('port'),
            'password': config.get('password'),
            'mount': config.get('mount', '/radio.mp3'),
            'name': config.get('name', 'MRadio'),
            'description': config.get('description', 'MRadio Stream'),
            'genre': config.get('genre', 'Various'),
            'bitrate': config.get('bitrate', '128'),
            'sampleRate': config.get('sampleRate', '44100'),
            'channels': config.get('channels', '2')
        }
        
        self.ffmpeg_process = None
        self.is_connected = False
        self.reconnect_attempts = 0
        self.max_reconnect_attempts = 10
        self.reconnect_delay = 5
        self.reconnect_timer = None
        self.is_reconnecting = False
        self.buffer = []
        self.buffer_size = 0
        self.max_buffer_size = 1024 * 1024
        self.connection_promise = None
    
    async def connect(self):
        if self.connection_promise:
            return await self.connection_promise
        
        if self.is_connected:
            logger.info('Already connected to Icecast server')
            return
        
        if self.is_reconnecting:
            logger.info('Already attempting to reconnect to Icecast server')
            return
        
        icecast_url = f"icecast://source:{self.config['password']}@{self.config['host']}:{self.config['port']}{self.config['mount']}"
        
        logger.info(f"Connecting to Icecast server at {self.config['host']}:{self.config['port']}{self.config['mount']}")
        
        ffmpeg_args = [
            get_ffmpeg_path(),
            '-hide_banner',
            '-loglevel', 'warning',
            '-f', 'mp3',
            '-i', 'pipe:0',
            '-err_detect', 'ignore_err',
            '-fflags', '+genpts+discardcorrupt',
            '-max_error_rate', '1.0',
            '-acodec', 'libmp3lame',
            '-ab', f"{self.config['bitrate']}k",
            '-ar', self.config['sampleRate'],
            '-ac', self.config['channels'],
            '-metadata', f"title={self.config['name']}",
            '-metadata', f"artist={self.config['description']}",
            '-metadata', f"genre={self.config['genre']}",
            '-f', 'mp3',
            '-content_type', 'audio/mpeg',
            '-ice_name', self.config['name'],
            '-ice_description', self.config['description'],
            '-ice_genre', self.config['genre'],
            '-ice_public', '1',
            icecast_url
        ]
        
        try:
            self.ffmpeg_process = subprocess.Popen(
                ffmpeg_args,
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE
            )
            
            self.flush_buffer()
            
            await asyncio.sleep(2)
            
            if self.ffmpeg_process and self.ffmpeg_process.poll() is None:
                self.is_connected = True
                self.is_reconnecting = False
                self.reconnect_attempts = 0
                logger.info('FFmpeg Icecast streaming initialized')
            
        except Exception as error:
            logger.error(f'FFmpeg process error: {str(error)}')
            if not self.is_reconnecting:
                await self.handle_connection_error()
    
    async def handle_connection_error(self):
        if self.is_reconnecting:
            return
        
        self.is_connected = False
        
        if self.reconnect_attempts < self.max_reconnect_attempts:
            await self.schedule_reconnect()
        else:
            logger.error('Max reconnection attempts reached for Icecast')
    
    async def schedule_reconnect(self):
        if self.is_reconnecting:
            return
        
        self.is_reconnecting = True
        self.reconnect_attempts += 1
        delay = min(self.reconnect_delay * self.reconnect_attempts, 30)
        
        logger.info(f"Scheduling Icecast reconnection attempt {self.reconnect_attempts}/{self.max_reconnect_attempts} in {delay}s")
        
        await asyncio.sleep(delay)
        
        self.is_reconnecting = False
        self.cleanup()
        
        try:
            await self.connect()
        except Exception as err:
            logger.error(f'Icecast reconnection failed: {str(err)}')
            self.is_reconnecting = False
    
    def flush_buffer(self):
        if len(self.buffer) > 0 and self.ffmpeg_process and self.ffmpeg_process.stdin:
            try:
                for chunk in self.buffer:
                    self.ffmpeg_process.stdin.write(chunk)
                self.ffmpeg_process.stdin.flush()
                self.buffer = []
                self.buffer_size = 0
                logger.debug('Flushed audio buffer to stream')
            except Exception as error:
                logger.error(f'Error flushing buffer: {str(error)}')
    
    def write(self, chunk: bytes):
        if not chunk or len(chunk) == 0:
            return
        
        if self.is_connected and self.ffmpeg_process and self.ffmpeg_process.stdin:
            try:
                if len(self.buffer) > 0:
                    self.flush_buffer()
                
                self.ffmpeg_process.stdin.write(chunk)
                self.ffmpeg_process.stdin.flush()
            except Exception as error:
                logger.error(f'Error writing to Icecast stream: {str(error)}')
                self.add_to_buffer(chunk)
                if not self.is_reconnecting:
                    asyncio.create_task(self.handle_connection_error())
        else:
            self.add_to_buffer(chunk)
            
            if not self.is_connected and not self.is_reconnecting and self.reconnect_attempts == 0:
                asyncio.create_task(self.connect())
    
    def add_to_buffer(self, chunk: bytes):
        if self.buffer_size + len(chunk) <= self.max_buffer_size:
            self.buffer.append(chunk)
            self.buffer_size += len(chunk)
        else:
            while self.buffer_size + len(chunk) > self.max_buffer_size and len(self.buffer) > 0:
                removed = self.buffer.pop(0)
                self.buffer_size -= len(removed)
            self.buffer.append(chunk)
            self.buffer_size += len(chunk)
    
    def cleanup(self):
        if self.ffmpeg_process:
            try:
                if self.ffmpeg_process.stdin:
                    self.ffmpeg_process.stdin.close()
                self.ffmpeg_process.terminate()
                self.ffmpeg_process.wait(timeout=5)
            except Exception as error:
                logger.debug(f'Error cleaning up FFmpeg process: {str(error)}')
            self.ffmpeg_process = None
    
    def disconnect(self):
        if self.reconnect_timer:
            self.reconnect_timer = None
        
        self.cleanup()
        
        self.is_connected = False
        self.is_reconnecting = False
        self.reconnect_attempts = 0
        self.buffer = []
        self.buffer_size = 0
        
        logger.info('Disconnected from Icecast server')
    
    def get_status(self):
        return {
            'connected': self.is_connected,
            'reconnecting': self.is_reconnecting,
            'reconnectAttempts': self.reconnect_attempts,
            'bufferSize': self.buffer_size,
            'config': {
                'host': self.config['host'],
                'port': self.config['port'],
                'mount': self.config['mount'],
                'name': self.config['name']
            }
        }
